# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: parser/html/nsHtml5TreeBuilderHSupplement.h
# Commit: 77388b0795b4
# Full Hash: 77388b0795b4f9a30171c922d6b649c52e5f6c87
# Author: Henri Sivonen <hsivonen@hsivonen.fi>
# Date: 2021-12-23 09:34:36
# Description:
#   Bug 1745142 - Communicate encoding commitment via speculative load queue. r=smaug
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D133996
# ==============================================================================

diff -r a388127095db -r 77388b0795b4 parser/html/nsHtml5TreeBuilderHSupplement.h
--- a/parser/html/nsHtml5TreeBuilderHSupplement.h	Wed Dec 22 15:47:08 2021 +0000
+++ b/parser/html/nsHtml5TreeBuilderHSupplement.h	Wed Dec 22 15:54:49 2021 +0000
@@ -25,6 +25,15 @@
 nsresult mBroken;
 bool mCurrentHtmlScriptIsAsyncOrDefer;
 bool mPreventScriptExecution;
+/**
+ * Whether to actually generate speculative load operations that actually
+ * represent speculative loads as opposed to other operations traveling
+ * in the same queue. True for normal loads and false for XHR, plain text,
+ * and View Source. Needed, because we can't just null-check
+ * mSpeculativeLoadStage, since it is used for transferring encoding
+ * information even in the XHR/plain text/View Source cases.
+ */
+bool mGenerateSpeculativeLoads;
 #ifdef DEBUG
 bool mActive;
 #endif
@@ -84,7 +93,8 @@
 public:
 explicit nsHtml5TreeBuilder(nsHtml5OplessBuilder* aBuilder);
 
-nsHtml5TreeBuilder(nsAHtml5TreeOpSink* aOpSink, nsHtml5TreeOpStage* aStage);
+nsHtml5TreeBuilder(nsAHtml5TreeOpSink* aOpSink, nsHtml5TreeOpStage* aStage,
+                   bool aGenerateSpeculativeLoads);
 
 ~nsHtml5TreeBuilder();
 
@@ -104,8 +114,21 @@
 
 void FlushLoads();
 
+/**
+ * Sets the document charset via the speculation queue.
+ *
+ * @param aCommitEncodingSpeculation true iff the main thread should
+ *                           treat the first speculation as an
+ *                           encoding speculation.
+ */
 void SetDocumentCharset(NotNull<const Encoding*> aEncoding,
-                        int32_t aCharsetSource);
+                        nsCharsetSource aCharsetSource,
+                        bool aCommitEncodingSpeculation);
+
+/**
+ * Updates the charset source via the op queue.
+ */
+void UpdateCharsetSource(nsCharsetSource aCharsetSource);
 
 void StreamEnded();
 